Quasi-static beam

sveerara opened the pull request:
This PR contains source files, tests and documentation files for Quasi-static beam element.

@acasagran, @bwspenc please review it when you have time. 

refs #10313

sveerara committed:
Quasi-static beam #10313

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/171837/) on 8bb4cf2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11148/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11148/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1ea4d0012b53f029401025d1cadac19a6982788e`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/171847/) on 399e814 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11148/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/171970/) on 217bbb3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11148/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11148/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 710036da7dfca4e1b663809fc55bdd0b00d62b3e`

acasagran commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,595 @@
+                               "Second moment of area of the beam about z axis. Can be "
+                               "supplied as either a number or a variable name.");
+  params.addParam<bool>(
+      "large_strain", false, "Set to true if large strain have to be calculated.");
```
if large **strains** ~~have~~ **are** to be calculated.

acasagran commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,595 @@
+  _total_rot_strain[_qp] = _total_rotation[0].transpose() * _mech_rot_strain_increment[_qp] +
+                           _total_disp_strain_old[_qp];
+
+  // Convert eigen strain increment from global to beam local coordinate system and remove eigen
```
~~eigen strain~~ **eigenstrain**

acasagran commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,595 @@
+
+    k1_large_22 *= 1.0 / 4.0 / Utility::pow<2>(_original_length[0]);
+
+    // k2_large and k3_large are constributions from tau_xy * d(gamma_xy) and tau_xz * d(gamma_xz)
```
~~constributions~~ **contributions**

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/euler_finite_rot_y.i:
```diff
@@ -0,0 +1,237 @@
+# Large strain/large rotation cantilever beam tese
```
~~tese~~ **test**

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/euler_finite_rot_y.i:
```diff
@@ -0,0 +1,237 @@
+# Area (A) = 1.0
+# Iy = Iz = 0.16
+
+# The non-dimensionless parameter alpha = kAGL^2/EI = 1e6
```
~~non-dimensionless~~ use either **non-dimensional** or **dimensionless**

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/euler_finite_rot_y.i:
```diff
@@ -0,0 +1,237 @@
+[Executioner]
+  type = Transient
+  solve_type = PJFNK
+#  petsc_options_iname = '-pc_type -ksp_gmres_restart'
```
Delete commented out lines here and below

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/euler_finite_rot_z.i:
```diff
@@ -0,0 +1,237 @@
+# Large strain/large rotation cantilever beam tese
```
~~tese~~ **test**

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/euler_finite_rot_z.i:
```diff
@@ -0,0 +1,237 @@
+# Area (A) = 1.0
+# Iy = Iz = 0.16
+
+# The non-dimensionless parameter alpha = kAGL^2/EI = 1e6
```
~~non-dimensionless~~ use either **non-dimensional** or **dimensionless**

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/euler_finite_rot_z.i:
```diff
@@ -0,0 +1,237 @@
+[Executioner]
+  type = Transient
+  solve_type = PJFNK
+#  petsc_options_iname = '-pc_type -ksp_gmres_restart'
```
Delete commented out lines

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/euler_small_strain_y.i:
```diff
@@ -0,0 +1,222 @@
+# The ratio beam FEM solution and analytical solution is 0.998.
+
+# References:
+# Prathap and Bashyam (1982), International journal for numerical methods in engineering, vol. 18, 195-210.
```
~~Bashyam~~ **Bhashyam**

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/euler_small_strain_y.i:
```diff
@@ -0,0 +1,222 @@
+[Executioner]
+  type = Transient
+  solve_type = NEWTON
+#  petsc_options_iname = '-pc_type -ksp_gmres_restart'
```
Delete commented out lines

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/euler_small_strain_z.i:
```diff
@@ -0,0 +1,222 @@
+# The ratio beam FEM solution and analytical solution is 0.998.
+
+# References:
+# Prathap and Bashyam (1982), International journal for numerical methods in engineering, vol. 18, 195-210.
```
~~Bashyam~~ **Bhashyam**

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/timoshenko_small_strain_y.i:
```diff
@@ -0,0 +1,222 @@
+# This deflection matches the FEM solution given in Prathap and Bashyam (1982).
+
+# References:
+# Prathap and Bashyam (1982), International journal for numerical methods in engineering, vol. 18, 195-210.
```
~~Bashyam~~ **Bhashyam**

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/timoshenko_small_strain_z.i:
```diff
@@ -0,0 +1,222 @@
+# This deflection matches the FEM solution given in Prathap and Bashyam (1982).
+
+# References:
+# Prathap and Bashyam (1982), International journal for numerical methods in engineering, vol. 18, 195-210.
```
~~Bashyam~~ **Bhashyam**

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/euler_small_strain_z.i:
```diff
@@ -0,0 +1,222 @@
+[Executioner]
+  type = Transient
+  solve_type = NEWTON
+#  petsc_options_iname = '-pc_type -ksp_gmres_restart'
```
Delete commented out lines

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/timoshenko_small_strain_y.i:
```diff
@@ -0,0 +1,222 @@
+[Executioner]
+  type = Transient
+  solve_type = NEWTON
+#  petsc_options_iname = '-pc_type -ksp_gmres_restart'
```
Delete commented out lines

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/timoshenko_small_strain_z.i:
```diff
@@ -0,0 +1,222 @@
+[Executioner]
+  type = Transient
+  solve_type = NEWTON
+#  petsc_options_iname = '-pc_type -ksp_gmres_restart'
```
Delete commented out lines

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/timoshenko_small_strain_z.i:
```diff
@@ -0,0 +1,222 @@
+# delta = PL^3/3EI * (1 + 3.0 / alpha) = 5.868e-2m
+
+# Using 10 elements to discretize the beam element, the FEM solution is 5.852e-2 m.
+# This deflection matches the FEM solution given in Prathap and Bashyam (1982).
```
~~Bashyam~~ **Bhashyam**

acasagran commented on modules/tensor_mechanics/test/tests/beam/static/timoshenko_small_strain_y.i:
```diff
@@ -0,0 +1,222 @@
+# delta = PL^3/3EI * (1 + 3.0 / alpha) = 5.868e-4 m
+
+# Using 10 elements to discretize the beam element, the FEM solution is 5.852e-2m.
+# This deflection matches the FEM solution given in Prathap and Bashyam (1982).
```
~~Bashyam~~ **Bhashyam**

acasagran reviewed (commented):
@sveerara We discussed adding support for using any 2 elastic constants to specify the beam properties. Has that been done? If so, should a test using say E and nu be added?

sveerara commented on modules/tensor_mechanics/doc/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md:
```diff
@@ -0,0 +1,156 @@
+{}^{t+\Delta t}R = \Delta R \; {}^tR
+$$
+
+The rotation matrix is incrementally updated if [ComputeFiniteBeamStrain](/ComputeFiniteBeamStrain.md) is used. If [ComputeIncrementalBeamStrain](/ComputeIncrementalBeamStrain.md) is used, the rotation matrix is same as the initially computed rotation matrix ($^0R$).
```
changed transformation matrix to rotation matrix throughout the file.

sveerara commented on modules/tensor_mechanics/doc/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md:
```diff
@@ -0,0 +1,156 @@
+^{t+\Delta t} {\{ {\Delta \theta} \}}^{local} = {}^tR \;\; ^{t+\Delta t} \{ {\Delta \theta} \}
+$$
+
+These beam local displacements are then used in the calculation of the Euler angles $\alpha$, $\beta$ and $\gamma$ as in [cite:bathe_large_1979]. These Euler angles are used to calculate the rotation increment $\Delta R$ using which the rotation matrix at $t+\Delta t$ can be calculated as:
```
Changed this to read `These Euler angles are used to calculate the rotation increment $\Delta R$. Using $\Delta R$, the rotation matrix at $t+\Delta t$ can be calculated as:`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/172173/) on 69e1ea2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11148/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11148/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 710036da7dfca4e1b663809fc55bdd0b00d62b3e`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/172262/) on e21f729 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11148/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11148/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 710036da7dfca4e1b663809fc55bdd0b00d62b3e`

sveerara commented:
@acasagran I have addressed your comments. In ComputeElasticityBeam.C, I have added poissons_ratio parameter and now any two of E, G and nu can be provided as input. I have modified the euler_pipe_axial_force test to check for all three combinations and I have added error tests when 2 parameters are not provided or when all three parameters are provided. Thanks!

sveerara commented:
@bwspenc, can you please look at this PR when you have time? I have addressed Al's comments on this PR.

cbolisetti commented on modules/tensor_mechanics/doc/content/documentation/modules/tensor_mechanics/C0TimoshenkoBeam.md:
```diff
@@ -0,0 +1,161 @@
+= \int_{0}^{{}^0L} F_1(x) \delta \epsilon_{1}(x) + F_2(x) \delta \epsilon_{2}(x) + F_3(x) \delta \epsilon_{3}(x) + M_1(x) \delta \kappa_1(x) + M_2(x) \delta \kappa_2(x) + M_3(x) \delta \kappa_3(x)
+\end{equation}
+
+As can be seen from the equation for $\epsilon_{1}(x)$, it depends only on $\frac{\partial u_1}{\partial x}$ if $A_y$ and $A_z$ are zero. But $\epsilon_{2}(x)$ depends on both $\frac{\partial u_1}{\partial x}$ and $\theta_3$. Therefore, $F_2(x) \delta \epsilon_{2}(x)$ would contribute to the residual of both the translational DOF in y direction and rotational DOF in the z direction. Similarly, $F ...
```
@sveerara I think you're missing the references section here. The bibliography is not showing up.

sveerara replied:
@cbolisetti, thanks for letting me know. I have added the bibliography line to the .md file. Surprisingly none of the tensor mechanics files have in documentation/modules/tensor_mechanics have a references section. Looks like the bibliography line somehow got dropped when it was moved to the new documentation system. 

Also, thanks for pointing out that `y_orientation` is not normalized within ComputeIncrementalBeamStrain.C. I have corrected that now.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/173699/) on e891c20 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11148/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11148/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 710036da7dfca4e1b663809fc55bdd0b00d62b3e`

sveerara commented on modules/tensor_mechanics/include/kernels/StressDivergenceBeam.h:
```diff
@@ -0,0 +1,109 @@
+  virtual Real computeQpResidual() override { return 0.0; }
+  virtual void computeJacobian() override;
+  virtual void computeOffDiagJacobian(MooseVariableFE & jvar) override;
+  using Kernel::computeOffDiagJacobian;
```
I tried after removing `using Kernel::computeOffDiagJacobian`. While compiling in debug mode, I get the following error

```
In file included from tensor_mechanics/build/header_symlinks/ALEKernel.h:13:0,
                 from /opt/civet/build_0/moose/modules/tensor_mechanics/src/kernels/ALEKernel.C:10,
                 from tensor_mechanics/build/unity_src/kernels_Unity.C:2:
../framework/build/header_symlinks/Kernel.h:38:16: error: 'virtual void Kernel::computeOffDiagJacobian(unsigned int)' was hidden [-Werror=overloaded-virtual]
   virtual void computeOffDiagJacobian(unsigned jvar);
                ^~~~~~~~~~~~~~~~~~~~~~
In file included from /opt/civet/build_0/moose/modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:8:0,
                 from tensor_mechanics/build/unity_src/kernels_Unity.C:14:
tensor_mechanics/build/header_symlinks/StressDivergenceBeam.h:28:16: error:   by 'virtual void StressDivergenceBeam::computeOffDiagJacobian(MooseVariableFE&)' [-Werror=overloaded-virtual]
   virtual void computeOffDiagJacobian(MooseVariableFE & jvar) override;
                ^~~~~~~~~~~~~~~~~~~~~~
```

I am not exactly sure what it means. But I am surprised why ALEKernel would generate an error when StressDivergenceBeam derives from Kernel and not ALEKernel.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/173773/) on a15a19f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11148/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11148/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 710036da7dfca4e1b663809fc55bdd0b00d62b3e`

sveerara commented on modules/tensor_mechanics/src/materials/ComputeFiniteBeamStrain.C:
```diff
@@ -0,0 +1,123 @@
+    dxyz(i) = (*node[1])(i) - (*node[0])(i);
+
+  Real original_length = dxyz.norm();
+
```
Since I moved the original length calculation to computeQpStrain() function in ComputeIncrementalBeamStrain.C, I had to calculate it again for the rotation calculation.

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+StressDivergenceBeam::computeResidual()
+{
+  DenseVector<Number> & re = _assembly.residualBlock(_var.number());
+  mooseAssert(re.size() == 2, "Beam element has and only has two nodes.");
```
The error message above doesn't quite make sense.

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+  _local_re.resize(re.size());
+  _local_re.zero();
+
+  RealVectorValue a(3, 0.0);
```
The size of RealVectorValue is hard-coded to 3. Just do this to get an empty size 3 vector:
`RealVectorValue a;`

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+
+  RealVectorValue a(3, 0.0);
+  std::vector<RealVectorValue> global_force_res(_test.size(), a);
+  std::vector<RealVectorValue> global_moment_res(_test.size(), a);
```
It's expensive to create these vectors every time this function gets called. Create these as member variables that get re-used instead.

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+
+  // add contributions from stiffness proportional damping (non-zero _zeta) or HHT time integration
+  // (non-zero _alpha)
+  if ((std::abs(_alpha) > 0.0 || _zeta[0] > 0.0) && _dt > 0.0)
```
Maybe do this check upfront once and have a boolean that tells you whether to compute dynamic terms.

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+  {
+    if (_component < 3)
+      _local_re(_i) = global_force_res[_i](_component);
+    else if (_component > 2)
```
this can just be an `else`

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+    for (unsigned int j = 0; j < _phi.size(); ++j)
+      if (_component < 3)
+        _local_ke(i, j) = (i == j ? 1 : -1) * _K11[0](_component, _component);
+      else if (_component > 2)
```
`else`

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+      }
+
+  // scaling factor for Rayliegh damping and HHT time integration
+  if ((std::abs(_alpha) > 0.0 || _zeta[0] > 0.0) && _dt > 0.0)
```
same as comment before about having a boolean member variable instead

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+    bool disp_coupled = false;
+    bool rot_coupled = false;
+
+    for (unsigned int i = 0; i < _ndisp; ++i)
```
Put curly brackets around these

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+    _local_ke.resize(ke.m(), ke.n());
+    _local_ke.zero();
+
+    if (disp_coupled || rot_coupled)
```
Put curly braces around these nested blocks

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+            else
+              _local_ke(i, j) += _K22_cross[0](_component - 3, coupled_component - 3);
+          }
+    else if (false) // Need some code here for coupling with temperature
```
Just remove these 3 lines

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+StressDivergenceBeam::computeDynamicTerms(std::vector<RealVectorValue> & global_force_res,
+                                          std::vector<RealVectorValue> & global_moment_res)
+{
+  RealVectorValue a(3, 0.0);
```
`RealVectorValue a;`

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+                                            std::vector<RealVectorValue> & global_moment_res)
+{
+  std::vector<RealVectorValue> force_local_t(_qrule->n_points());
+  std::vector<RealVectorValue> moment_local_t(_qrule->n_points());
```
Have these be member variables, and call `resize()` here.

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergenceBeam.C:
```diff
@@ -0,0 +1,319 @@
+
+  // residual for displacement variables
+  std::vector<RealVectorValue> local_force_res(_test.size());
+  for (_i = 0; _i < _test.size(); ++_i)
```
curly braces

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeElasticityBeam.C:
```diff
@@ -0,0 +1,70 @@
+void
+ComputeElasticityBeam::computeQpProperties()
+{
+  Real shear_modulus = _youngs_modulus[_qp] / (2.0 * (1.0 + _poissons_ratio[_qp]));
```
Make this `const`

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeFiniteBeamStrain.C:
```diff
@@ -0,0 +1,123 @@
+
+  RealVectorValue disp0(3, 0.0), disp1(3, 0.0);
+  RealVectorValue rot0(3, 0.0), rot1(3, 0.0);
+  RealVectorValue delta_disp(3, 0.0);
```
`RealVectorValue disp0, disp1, rot0, rot1, delta_disp;`

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeFiniteBeamStrain.C:
```diff
@@ -0,0 +1,123 @@
+  RealVectorValue rot0(3, 0.0), rot1(3, 0.0);
+  RealVectorValue delta_disp(3, 0.0);
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
```
Since you're reusing these indices so many times here, probably better to create a local variable that is set once and re-used many times.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeFiniteBeamStrain.C:
```diff
@@ -0,0 +1,123 @@
+  RealVectorValue delta_disp_local(_total_rotation_old[0] * delta_disp);
+
+  // Third - calculate the incremental rotation matrix using Euler angles
+  Real intermediate_length_1 = std::sqrt(Utility::pow<2>(original_length + delta_disp_local(0)) +
```
A lot of these can be `const`

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeFiniteBeamStrain.C:
```diff
@@ -0,0 +1,123 @@
+             rotation_d_1(2) * avg_rot_local(2));
+
+  RankTwoTensor rotation_a;
+  rotation_a.zero();
```
This isn't necessary because it is initialized with zeros.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,590 @@
+    _K22(declareProperty<RankTwoTensor>("Jacobian_22")),
+    _K22_cross(declareProperty<RankTwoTensor>("Jacobian_22_cross")),
+    _large_strain(getParam<bool>("large_strain")),
+    _grad_disp_0_local_t(3, 0.0),
```
Don't pass anything into the constructors of these `RealVectorValue` variables.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,590 @@
+  : Material(parameters),
+    _nrot(coupledComponents("rotations")),
+    _ndisp(coupledComponents("displacements")),
+    _rot_num(3),
```
`_rot_num(_nrot)`?

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,590 @@
+ComputeIncrementalBeamStrain::ComputeIncrementalBeamStrain(const InputParameters & parameters)
+  : Material(parameters),
+    _nrot(coupledComponents("rotations")),
+    _ndisp(coupledComponents("displacements")),
```
Check to make sure displacements and rotations are the same size.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,590 @@
+  // global to original beam local configuration and is never updated
+  _total_rotation[_qp] = _original_local_config[_qp];
+
+  _K11[0].zero();
```
A lot of non-stateful properties are being set here, which is a problem because they'll be lost.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,590 @@
+  const NumericVector<Number> & sol = *nonlinear_sys.currentSolution();
+  const NumericVector<Number> & sol_old = nonlinear_sys.solutionOld();
+
+  RealVectorValue disp0(3, 0.0), disp1(3, 0.0);
```
Remove arguments to these constructors.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,590 @@
+  RealVectorValue delta_disp(3, 0.0);
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
+    disp0(i) = sol(node[0]->dof_number(nonlinear_sys.number(), _disp_num[i], 0)) -
```
Store off these indices and reuse them.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,590 @@
+  // rot_strain_3 = integral(e_11 * y) dA
+  // J is the product moment of inertia which is zero for most cross-sections so it is assumed to be
+  // zero for this analysis
+  Real J = 0;
```
Make J const

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,590 @@
+
+void
+ComputeIncrementalBeamStrain::computeJacobian()
+{
```
This is analogous to the material stiffness calculated by the stress calculator. Could this be moved to `computeBeamResultants`? Also, I'd call it something other than `computeJacobian` to avoid confusion with the Kernel's `computeJacobian`.

sveerara replied:
For the small strain scenario, the stiffness matrix depends only on the material properties and the geometry of the beam. But for the large strain scenario, the displacements of the beam are also needed to get the stiffness matrix. So if I move the stiffness matrix calculation to ComputeBeamResultants, I will have either redo some of the displacement calculation or transfer those over from the ComputeIncrementalBeamStrain to ComputeBeamResultants.

bwspenc replied:
I think this is pretty analogous to what happens in `StressDivergenceTensors::computeFiniteDeformJacobian()`.  I had suggested that the stiffness be computed in `ComputeBeamResultants`, but I guess it would really need to be moved to the Kernel to be consistent with that.  Aside from this, I think this code is ready to go, and I'm going to merge it.

I would however still like to look into whether we can make this more consistent with how the solid mechanics equivalent works. To avoid re-computing displacements or other quantities, maybe we can just make them material properties that the Kernel accesses.

sveerara replied:
Thanks, Ben. I will edit and submit the other PRs and when you and Al are reviewing the other PRs, I will look into whether this can be made more consistent with the solid mechanics equivalent.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeIncrementalBeamStrain.C:
```diff
@@ -0,0 +1,590 @@
+  Real youngs_modulus = _material_stiffness[0](0);
+  Real shear_modulus = _material_stiffness[0](1);
+
+  Real A_avg = (_area[0] + _area[1]) / 2.0;
```
Make these `const`

sveerara committed:
Addressing comments #10313

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/178457/) on 46bb199 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11148/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11148/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5962d4dc56578db13bdd96febf454faf5f272834`

sveerara commented:
@bwspenc, I tried moving the nonQp functions (ComputeResidual, ComputeJacobian and ComputeOffDiagJacobian) to public from protected. That dint help with the error.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/178479/) on c8c248a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11148/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11148/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5962d4dc56578db13bdd96febf454faf5f272834`
